src/optimizers.jl

```diff
@@ -1,239 +1,147 @@
+      l = commonind(M[1],M[2]) 
+      return (i,o,l,ξ)
+    elseif site == length(M)
+      i = firstind(M[site], tags="Input", plev = 0) 
+      o = firstind(M[site], tags="Output", plev = 0) 
+      ξ = firstind(M[site], tags="Purifier") 
+      l = commonind(M[site],M[site-1])
+      return (i,o,l,ξ)
+    else
+      i = firstind(M[site], tags="Input", plev = 0) 
+      o = firstind(M[site], tags="Output", plev = 0) 
+      ξ = firstind(M[site], tags="Purifier") 
+      l1 = commonind(M[site],M[site-1])
+      l2 = commonind(M[site],M[site+1])
+      return (i,o,l1,l2,ξ)
+    end
+  else
+    if site == 1
+      s = firstind(M[site], tags="Site") 
+      ξ = firstind(M[site], tags="Purifier") 
+      l = commonind(M[1],M[2]) 
+      return (s,l,ξ)
+    elseif site == length(M)
+      s = firstind(M[site], tags="Site") 
+      ξ = firstind(M[site], tags="Purifier") 
+      l = commonind(M[site],M[site-1])
+      return (s,l,ξ)
+    else
+      s = firstind(M[site], tags="Site") 
+      ξ = firstind(M[site], tags="Purifier") 
+      l1 = commonind(M[site],M[site-1])
+      l2 = commonind(M[site],M[site+1])
+      return (s,l1,l2,ξ)
     end
   end
-  for j in 1:length(M)
-    opt.∇²[j] += abs2.(∇[j])
-    ∇² = copy(opt.∇²[j])
-    ∇² .+= opt.ϵ
-    g = sqrt.(∇²)
-    δ = g .^ -1
-    M[j] = M[j] - opt.η * (noprime(∇[j]) ⊙ δ)
-  end
-end
-
-update!(ψ::MPS, ∇::Array, opt::AdaGrad; kwargs...) = update!(LPDO(ψ), ∇, opt; kwargs...)
-
-struct AdaDelta <: Optimizer
-  γ::Float64
-  ϵ::Float64
-  ∇²::Vector{ITensor}
-  Δθ²::Vector{ITensor}
 end
 
-Base.copy(opt::AdaDelta) = AdaDelta(opt.γ, opt.ϵ, copy(opt.∇²), copy(opt.Δθ²))
-
-"""
-    AdaDelta(L::LPDO;γ::Float64=0.9,ϵ::Float64=1E-8)
-
-# Parameters
-  - `γ`: friction coefficient
-  - `ϵ`: shift 
-  - `∇²`: square gradients (decaying average)
-  - `Δθ²`: square updates (decaying average)
-"""
-function AdaDelta(; γ::Float64=0.9, ϵ::Float64=1E-8)
-  Δθ² = ITensor[]
-  ∇² = ITensor[]
-  return AdaDelta(γ, ϵ, ∇², Δθ²)
-end
-
-"""
-    update!(L::LPDO,∇::Array,opt::AdaDelta; kwargs...)
-    
-    update!(ψ::MPS,∇::Array,opt::AdaDelta; kwargs...)
-
-Update parameters with AdaDelta
-
-
-1. `gⱼ = γ * gⱼ + (1-γ) * ∇ⱼ²`: decaying average
-2. `Δθⱼ = ∇ⱼ * sqrt(pⱼ) / sqrt(gⱼ+ϵ) ` 
-3. `θⱼ = θⱼ - Δθⱼ`: parameter update
-4. `pⱼ = γ * pⱼ + (1-γ) * Δθⱼ²`: decaying average
-"""
-function update!(L::LPDO, ∇::Array, opt::AdaDelta; kwargs...)
-  M = L.X
-  if isempty(opt.∇²)
-    for j in 1:length(M)
-      push!(opt.Δθ², ITensor(zeros(size(M[j])), inds(M[j])))
-      push!(opt.∇², ITensor(zeros(size(M[j])), inds(M[j])))
+function getparameters(L::LPDO; ∇ = nothing)
+  X = L.X
+  N = length(X)
+  θ = Complex[]
+  λ = Complex[]
+  # either state or gradients
+  for j in 1:N
+    legs = gettensorindices(X, j)
+    basis = [collect(1:x) for x in dim.(legs)]
+    basis = vec(Iterators.product(basis...)|>collect)
+    for μ in basis
+      tels = legs .=> μ
+      push!(θ, X[j][tels...])
```

Comment:
It took me a bit to follow what this was doing. Are you saving the values of the ITensors element by element? I think that will be very slow.

Reply:
Yeah, the other option would be the reshuffle the indices so they match a given order and just do `array(M[site])` but I am unsure on how to do that.

Reply:
Can't you just do something like `collect(Iterators.flatten(array.(L)))` (i.e. a flattened representation of the data of the ITensors), pass that vector to Optimisers, and then read the data back using the indices of `L`? Maybe I'm missing something about the requirements about the data ordering.

Reply:
But the order of the flattened `L` elements must match the order of the flattened gradient. So one has to make sure the gradient index ordering is the same as `L`.

Reply:
When I say index ordering I mean something like `("Site","Link","Link")` vs `("Link","Site","Link")` for a given tensor.

Reply:
Right, but can't you just use the index ordering from `L` when you read back the data?